Fixes #38 - Update overview & migrate 2 articles from source (#39)
MicrosoftDocs/PowerShell-Docs-Modules: reference/docs-conceptual/PSScriptAnalyzer/rules-recommendations.md

---
description: This article lists best-practice recommendations and the rules associated with them.
ms.date: 03/22/2022
title: PSScriptAnalyzer rules and recommendations
---
# PSScriptAnalyzer rules and recommendations

The following guidelines come from a combined effort from both the PowerShell team and the
community. The guidelines are organized by type. Within each type there is a list of rules. The

> rules are group by the **Severity** defined in the implementation of the **PSScriptAnalyzer** rule.

Reviewer: ```suggestion
rules are grouped by the **Severity** defined in the implementation of the **PSScriptAnalyzer** rule.
```

  [AvoidReservedParams](Rules/ReservedParams.md)
- Support Confirmation Requests
  [UseShouldProcessForStateChangingFunctions](Rules/UseShouldProcessForStateChangingFunctions.md)
  and
  [UseShouldProcessForStateChangingFunctions](Rules/UseShouldProcessForStateChangingFunctions.md)
- Must call ShouldProcess when ShouldProcess attribute is present and vice
  versa.[UseShouldProcess](Rules/ShouldProcess.md)
- Nouns should be singular
  [UseSingularNouns](Rules/UseSingularNouns.md)
- Module Manifest Fields
  [MissingModuleManifestField](Rules/MissingModuleManifestField.md)
  - Version
  - Author
  - Description
  - LicenseUri (for PowerShell Gallery)
- Switch parameters should not default to true
  [AvoidDefaultValueSwitchParameter](Rules/AvoidDefaultValueSwitchParameter.md)

### Severity: Information

No rules defined.

### Severity: TBD

- Support Force Parameter for Interactive Session
- If your cmdlet is used interactively, always provide a Force parameter to override the interactive
  actions, such as prompts or reading lines of input). This is important because it allows your
  cmdlet to be used in non-interactive scripts and hosts. The following methods can be implemented

>   by an interactive host.

Reviewer: Should this be a sub-bullet of the entry above or part of it? These seem to go together.

  [UseDeclaredVarsMoreThanAssignments](Rules/UseDeclaredVarsMoreThanAssignments.md)
- No Invoke-Expression
  [AvoidUsingInvokeExpression](Rules/AvoidUsingInvokeExpression.md)

### Severity: Information

No rules defined.

### Severity: TBD

- `Clear-Host` should not be used
- File paths should not be used (UNC)
- Error Handling
  - Use `-ErrorAction Stop` when calling cmdlets
  - Use $ErrorActionPreference = 'Stop'/' Continue' when calling non-cmdlets
  - Avoid using flags to handle errors
  - Avoid using `$?`
  - Avoid testing for a null variable as an error condition
  - Copy `$Error[0]` to your own variable
- Avoid using pipelines in scripts
- If a return type is declared, the cmdlet must return that type. If a type is returned, a return
  type must be declared.

## Scripting Style

### Severity: Error

No rules defined.

### Severity: Warning

- Don't use `Write-Host` unless writing to the host is all you want to do
  [AvoidUsingWriteHost](Rules/AvoidUsingWriteHost.md)

### Severity: Information

- Write comment-based help
  [ProvideCommentHelp](Rules/ProvideCommentHelp.md)

### Severity: TBD

- Provide usage Examples

> - Use the Notes section for detail on how the tool work

Reviewer: ```suggestion
- Use the Notes section for details on how the tool works
```

- Should have help on every exported command (including parameter documentation)

> - Document the version of PowerShell that script was written for

Reviewer: ```suggestion
- Document the version of PowerShell that the script was written for
```

- Indent your code
- Avoid backticks

## Script Security

### Severity: Error

- Password should be secure string

>   [AvoidUsingPlainTextForPassword](Rules/AvoidUsingPlainTextForPassword.md) - Should never have both
>   -Username and -Password parameters (should take credentials):
>   [UsePSCredentialType](Rules/UsePSCredentialType.md)

Reviewer: ```suggestion
  [AvoidUsingPlainTextForPassword](Rules/AvoidUsingPlainTextForPassword.md)
- Should never have both `-Username` and `-Password` parameters (should take credentials):
  [UsePSCredentialType](Rules/UsePSCredentialType.md)
```
